### ISSUE TITLE
question:How to convert fine tuned model to mlx format ?

### ISSUE THREAD

**Issue by x4080** [lucasnewman/f5-tts-mlx#22]
Hi, How to convert fine tuned model to mlx format ?

Thanks

**Comment by tikikun** [comment]
same question

**Comment by x4080** [comment]
@tikikun I asked in the discussion on mlx github repo basically the same thing, tried searching using google, perplexity etc, it seems kinda mystery to convert the models hope @lucasnewman  can give us some enlightment

**Comment by enessusan00** [comment]
Same here

**Comment by postacik** [comment]
I am also looking for a way to convert a model to mlx format.

**Comment by lucasnewman** [comment]
Here's how I did it originally loading from a torch checkpoint. The alternative would be to convert the torch model to .safetensors first if you want to avoid the dependency:

```
import torch
import mlx.core as mx

ckpt_path = Path("model_1200000.pt")
state_dict = torch.load(ckpt_path, map_location='cpu', weights_only=True)['ema_model_state_dict']

# load weights
new_state_dict = {}
for k, v in state_dict.items():
    k = k.replace('ema_model.', '')
    v = mx.array(v.numpy())
    
    # rename layers
    if len(k) < 1 or 'mel_spec.' in k or k in ('initted', 'step'):
        continue
    elif '.to_out' in k:
        k = k.replace('.to_out', '.to_out.layers')
    elif '.text_blocks' in k:
        k = k.replace('.text_blocks', '.text_blocks.layers')
    elif '.ff.ff.0.0' in k:
        k = k.replace('.ff.ff.0.0', '.ff.ff.layers.0.layers.0')
    elif '.ff.ff.2' in k:
        k = k.replace('.ff.ff.2', '.ff.ff.layers.2')
    elif '.time_mlp' in k:
        k = k.replace('.time_mlp', '.time_mlp.layers')
    elif '.conv1d' in k:
        k = k.replace('.conv1d', '.conv1d.layers')
    
    # reshape weights
    if '.dwconv.weight' in k:
        v = v.swapaxes(1, 2)
    elif '.conv1d.layers.0.weight' in k:
        v = v.swapaxes(1, 2)
    elif '.conv1d.layers.2.weight' in k:
        v = v.swapaxes(1, 2)
    
    new_state_dict[k] = v

model.load_weights(list(new_state_dict.items()))
model.eval()
```

**Comment by x4080** [comment]
@lucasnewman Thanks for sharing, is there any reference about converting model to MLX ?

**Comment by postacik** [comment]
I tried to use the following model... 

https://huggingface.co/marduk-ra/F5-TTS-Turkish

...by using the following parameter...

--model "marduk-ra/F5-TTS-Turkish"

...but it failed probably because it has to be converted to be compatible with mlx.

What is the proper way of using torch compatible models with ft-tts-mlx?

**Comment by lucasnewman** [comment]
> I tried to use the following model...
> 
> https://huggingface.co/marduk-ra/F5-TTS-Turkish
> 
> ...by using the following parameter...
> 
> --model "marduk-ra/F5-TTS-Turkish"
> 
> ...but it failed probably because it has to be converted to be compatible with mlx.
> 
> What is the proper way of using torch compatible models with ft-tts-mlx?

You'll need to have a file named `model.safetensors` in the HF repository or it won't be able to find the model to load. I'll add an option to convert weights from the torch format to make it easier to use out of the box.

**Comment by postacik** [comment]
@lucasnewman hi,

Thank you very much for the update. 

I tried the 0.1.8 version. I had to change the following line in generate.py...

`f5tts = F5TTS.from_pretrained(model_name, True)
`
...since it does not include a command line option to convert weights, yet.

I also had to rename one of the existing .safetensors file to model.safetensors.

After these modifications it worked with the custom model.
